### PR TITLE
Define process for adding games over network

### PR DESCRIPTION
**Pull request by Wills2022**
# Description
Company leadership has identified a requirement to be able to add games to the arcade machines over a local network connection.  This is to facilitate an upcoming Hackathon occurring over the 2024 T2 Intra-trimester break. 

The requirements outline a process by which students can upload their games to an Arcade Machine. The process I have outlined offers an SFTP and SCP approach that most students should be able to utilize. The guide has been written with the Thoth-Tech Arcade Machine implementation in mind (in regards to variables and steps). 

The guide assumes that students are A) able to cross-compile their games to arm32 binaries or B) provide a arm32 .deb file with which to install the games from.

## Type of change
- [X] Documentation (New)

# How Has This Been Tested?
Tested on my local Raspberry PI and home network. My Raspberry Pi has been configured using the Thoth-Tech  Arcade Machine setup guide and should very closely mimic the actual configuration in use. 

Using this guide I am able to copy files to the RPI via SCP and SFTP using the WiFi hotspot running from the RPI.

## Testing Checklist:
- [X] Tested in latest Chrome
- [X] Tested in latest Firefox

# Checklist:
- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have requested a review from ... on the Pull Request

### CODE REVIEW

**Review by Liquidscroll — APPROVED** [review]
Looks good to me! Very thorough, leverages both the existing access points and uses common utilities.

File: docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
+# Adding Games To Arcade Machine
+
+## Pre Requisites:
+### WiFi Access Point
+This guide assumes that there is an existing WiFi Access Point on the Arcade Machine you are trying to copy games to.  You should be able to verify this by looking for nearby WiFi networks called “Arcade`X`” where `X` is the number of the Arcade Machine. 
+
+If you are unable to see the WiFi network, or you need to set it up follow this guide: [link](https://github.com/thoth-tech/documentation/blob/main/docs/Splashkit/Applications/Arcade%20Machines/Arcade%20Machine%20Setup/Setup%20Arcade%20Machine.md#setup-wifi-access-point-optional).
+
+### SSH Utilities
+In order transfer the files from your local machine to the Arcade Machine you will need to use either SCP or SFTP. Both applications are part of the SSH suite and come pre-installed on most modern operating systems. In order to test that these are installed open a terminal (If you're running **Windows** make sure you use **PowerShell**!). From the terminal window type ```scp``` and/or ```sftp```. You should output similar to the below prompting usage of these commands. If you do you're good to go!
```
Comment (hugglesfox) [inline docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md:10]: SFTP and SCP have rather different interfaces (as shown below) however the example below will only apply to scp which could cause confusion if you used sftp instead.

Additionally I don't see a need for typing both scp and sftp as you'll just end up with an invalid command. Hence I would suggest rewording this bit to say something like "In the terminal type either `scp` or `sftp`, you should get an output similar to the respective example below"

Comment (hugglesfox) [inline docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md:10]: Also a nit pick but use ` for inline code, ``` should be used for block code

File: docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+user@host:~$ scp
+usage: scp [-c cipher] [-D sftp_server_path] [-F ssh_config]
+           [-i identity_file] [-J destination] [-l limit] [-o ssh_option]
+           [-P port] [-S program] [-X sftp_option] source ... target
+user@host:~$
+```
+If you have trouble running these commands please consult the relevant documentation on installing ssh **client** software on your operating system.
+
+### Game Files
+This guide assumes that you are transferring **compiled** game binaries to the Arcade Machine. The Arcade Machine uses a Raspberry Pi, which has an ARM processor. When compiling your games make sure you **cross compile** for arm32. If your game is packaged as a .deb file you can install it connecting to the Arcade Machine via ssh (```ssh deakin@10.42.0.1```) and then running ```sudo dpkg -i myGame.deb``` to install it locally.
```
Comment (hugglesfox) [inline docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md:21]: Do we have a way of packaging a game as a deb?

File: docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+64 bytes from 10.42.0.1: icmp_seq=1 ttl=58 time=43.0 ms
+64 bytes from 10.42.0.1: icmp_seq=2 ttl=58 time=47.0 ms
+64 bytes from 10.42.0.1: icmp_seq=3 ttl=58 time=40.8 ms
+64 bytes from 10.42.0.1: icmp_seq=4 ttl=58 time=41.1 ms
+64 bytes from 10.42.0.1: icmp_seq=5 ttl=58 time=40.6 ms
+^C
+--- 10.42.0.1 ping statistics ---
+5 packets transmitted, 5 received, 0% packet loss, time 4001ms
+user@host:~$ 
+```
```
Comment (hugglesfox) [inline docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md:62]: This all seems very complex from a users point of view and there's the potential that someone doesn't know how to set a static ip address for their operating system (e.g. the process of setting a static ip on windows is painfully long). I'm wondering if it might be easier if we run `dnsmasq` on the pi's to provide a dhcp server?

File: docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+## Secure Copy Protocol (SCP)
+The simplest method of transferring files to the Arcade Machine is to use [SCP](https://en.wikipedia.org/wiki/Secure_copy_protocol). The basic syntax of this command is:
+
+```scp [options] [[user@]host1:]source [[user@]host2:]destination```
+
+In this scenario the source is your computer and the destination is the Arcade Machine. SCP requires you to authenticate with the destination host in order to send files over the network. This means that you need to provide authentication credentials for the **Arcade Machine**. By default the SSH user's credentials are ```deakin``` and ```qwerty```.
```
Comment (hugglesfox) [inline docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md:69]: When should one use the deakin user vs the qwerty user? Also what's the qwerty user used for?

**Review by hugglesfox — CHANGES_REQUESTED** [review]
Looks good, just a few questions and suggestions.

My personal preference would be for the guide to choose either scp or sftp to simplify it and avoid potential confusion. SCP does have the much nicer user interface in terms of "copy this command and you're done" factor, though it is (very) [slowly being deprecated](https://lwn.net/Articles/835962/). 

I'll leave you to make the call on whether we should recommend to use scp or sftp or whether the guide should continue including both.

One other thing is the wifi access point. The [arcade machine setup guide does say to connect to eduroam](https://github.com/thoth-tech/documentation/blob/main/docs/Splashkit/Applications/Arcade%20Machines/Arcade%20Machine%20Setup/Setup%20Arcade%20Machine.md#connect-to-eduroam), maybe we could access them over eduroam rather then them having their own access point (avoids us having to setup a dhcp server or use static ip addresses). Is there a way for one to get the ip address of the arcade machine through emulation station?

File: docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md (hunk 3 above)
Comment (Wills2022) [reply docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md:62]: This is a catch-22. On one hand setting a static IP address is an annoying process for some users, on the other hand we'd have to make _another_ configuration change to the Arcade Machine implement a DHCP server.  I'm personally disinclined adding another service to manage if we can at all avoid it, even if this comes at the cost of some user friendliness.

There's also a small amount of system resources overhead required to run the DHCP server which i'd rather not invest on the already constrained hardware.

I know that dnsmasq can be configured to operate in proxy mode, but we'd also need to make sure this doesn't interfere with how the devices operate on the eduroam network. 

Ultimately I don't think that setting static IP addresses should be outside the realm of capability of a user base that is consisting primarily of computer science students.

File: docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md (hunk 4 above)
Comment (Wills2022) [reply docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md:69]: Fair call. I've modified the language to be a bit more explicit.

File: docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md (hunk 1 above)
Comment (Wills2022) [reply docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md:10]: Fair enough, I've amended this throughout the document.

Comment (Wills2022) [reply docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md:10]: When I was testing this out just typing "sftp" without any arguments won't load the program, it just prints out usage syntax like the SCP example.  

The intention was that users would type "scp", then enter then "sftp" then enter to test both programs, but i see how the sentence "....`scp` and/or `sftp`.' could be interpreted ambiguously.  I've reworded it slightly to make it more clear.

**Comment by Wills2022** [comment]
> I'll leave you to make the call on whether we should recommend to use scp or sftp or whether the guide should continue including both.
>
I  disagree that giving users the option of either command is confusing. I hope that he document is written concisely & clearly enough to convey to users that they have the option of using whichever utility they find easier to use.  If forced my preference would be to retain the SCP section, as this is - in my opinion - the simpler tool to use. 
 
> One other thing is the wifi access point. The [arcade machine setup guide does say to connect to eduroam](https://github.com/thoth-tech/documentation/blob/main/docs/Splashkit/Applications/Arcade%20Machines/Arcade%20Machine%20Setup/Setup%20Arcade%20Machine.md#connect-to-eduroam), maybe we could access them over eduroam rather then them having their own access point (avoids us having to setup a dhcp server or use static ip addresses). Is there a way for one to get the ip address of the arcade machine through emulation station?
>
using eduroam is a can of worms I would prefer not to open.  From my perspective it is a network we do not "own", so it is very difficult for me to write comprehensive documentation that will not be invalidated due to infrastructure changes outside of our control. Namely the firewall rules and zoning.  I do not know whether Multicast DNS functions correctly on eduroam, which means we'd either need to have network administrators give us a static IP or fiddle around  getting the IP out of retropie somehow. 

In other words, There are too many variables (for my liking). i'd rather opt for a more complicated solution if it means we get ownership over the constituent pieces.

File: docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md (hunk 3 above)
Comment (hugglesfox) [reply docs/Splashkit/Applications/Arcade Machines/Arcade Machine Setup/AddGameOverNetwork.md]: Yeap that's fair.

**Review by hugglesfox — APPROVED** [review]
Yeah sounds fair, if we don't minimize complexity then the guide will be really blown out.

Thanks, looks good!

**Review by WhyPenguins — COMMENTED** [review]
Looks good overall! I agree that needing to setup a static IP isn't ideal, but it's reasonable considering the current setup. 

I'm happy to consider this complete, but just had a couple of questions after looking at `Adding Games.md`. I don't know much about the arcade machine's setup, so perhaps these questions don't make sense haha.

1. Is LaunchScripts where the games themselves should be copied? `Adding Games.md` says that LaunchScript (without the s) should contain shell scripts that launch the games, and the games are stored elsewhere. Perhaps `Adding Games.md` is out of date? Otherwise, how does Emulation Station know which file to execute?
2. After copying the games, will Emulation Station automatically pick them up on restart? Based on `Adding Games.md` they have to be added to `gamelist.xml`, so just copying the folder into LaunchScripts wouldn't be enough. I _think_ this makes sense, otherwise I'm unsure where the game's metadata comes from.

I presume these instructions were tested so I'm wondering if `Adding Games.md` is out of date or otherwise wrong, just wanted to check!

**Comment by WhyPenguins** [comment]
> Looks good overall! I agree that needing to setup a static IP isn't ideal, but it's reasonable considering the current setup.
> 
> I'm happy to consider this complete, but just had a couple of questions after looking at `Adding Games.md`. I don't know much about the arcade machine's setup, so perhaps these questions don't make sense haha.
> 
>     1. Is LaunchScripts where the games themselves should be copied? `Adding Games.md` says that LaunchScript (without the s) should contain shell scripts that launch the games, and the games are stored elsewhere. Perhaps `Adding Games.md` is out of date? Otherwise, how does Emulation Station know which file to execute?
> 
>     2. After copying the games, will Emulation Station automatically pick them up on restart? Based on `Adding Games.md` they have to be added to `gamelist.xml`, so just copying the folder into LaunchScripts wouldn't be enough. I _think_ this makes sense, otherwise I'm unsure where the game's metadata comes from.
> 
> 
> I presume these instructions were tested so I'm wondering if `Adding Games.md` is out of date or otherwise wrong, just wanted to check!

Sorry I'm pretty sure we already resolved these questions, but mind just writing briefly what the answers were in here? (I would if my memory wasn't so bad ahaha 😅) And yep just need to fix the linter complaints, otherwise looks good!

**Comment by Wills2022** [comment]
Howdy, my bad, could have sworn i updated a comment in here to address those questions. Based on my testing emulationstation will pick up on new games once the process has been restarted, but I couldn't find a way to have it dynamically re-scan the games/launcherscripts folders at run time and add new games in. For reference, the test menu directly launches new games that the user builds, rather than launching them via emulationstation. 

I've updated the file so that the automated checks should now run.
